### PR TITLE
implement Forward headers

### LINKED ISSUE #113: Set X-Forwarded-For header when forwarding HTTP requests

### LINKED ISSUE #155: Look into whether it makes sense to set X-Frame-Options
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options

### PR DESCRIPTION
**Pull request by jakobwgnr**
Closes #113 
Closes #155

### CODE REVIEW

**Comment by jakobwgnr** [comment]
Hello @kevinbader , @mmacai 
As discussed, the build is currently failing in `docker build` during `mix release`, due to some other issue not related to this PR.

As aligned, a new issue will be created for this and this issue should be fixed first.

== Compilation error in file lib/rig_inbound_gateway/api_proxy/router.ex ==
** (ArgumentError) can't fetch value for key `RigInboundGatewayWeb.Endpoint` of application `:rig_inbound_gateway`, can not resolve key HTTPS_CERTFILE value via adapter Elixir.Confex.Adapters.SystemEnvironment
    lib/confex.ex:155: Confex.fetch_env!/2
    lib/rig_inbound_gateway/api_proxy/router.ex:21: (module)
    (stdlib) erl_eval.erl:680: :erl_eval.do_apply/6

File: apps/rig_inbound_gateway/test/rig/api_proxy/router_test.exs [hunk 1]
```diff
@@ -174,6 +174,17 @@ defmodule RigInboundGateway.ApiProxy.RouterTest do
...
 
+  test_with_server "Any request should include forward headers", @env do
+    route("/myapi/direct", Response.ok!(~s<{"status":"ok"}>))
+
+    conn = call(Router, build_conn(:get, "/myapi/direct"))
+
+    assert conn.status == 200
+    assert conn.resp_body =~ "{\"status\":\"ok\"}"
+    assert get_req_header(conn, "X-Content-Type-Options") == ["nosniff"]
+    assert get_req_header(conn, "Forwarded") != [""]
```
Comment (kevinbader) [inline apps/rig_inbound_gateway/test/rig/api_proxy/router_test.exs]:
```suggestion
    assert [forwarded_header] = get_req_header(conn, "Forwarded")
    assert forwarded_header =~ ~r/for=[^;]+;by=.*/
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -9,9 +9,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- The proxy configuration can now also be passed as a JSON string. This allows to run the Docker image in environments where mounting a file in a container is not possible. [#159](https://github.com/Accenture/reactive-interaction-gateway/issues/159)
+- When using the proxy, RIG will now add additional forward handlers "X-Content-Type-Options" and "Forwarded" [#113](https://github.com/Accenture/reactive-interaction-gateway/issues/113)
 - Increased length of header value in HTTP requests to 16384. Reason is to support long tokens such as SAML.
 
 ### Changed
+- The proxy configuration can now also be passed as a JSON string. This allows to run the Docker image in environments where mounting a file in a container is not possible. [#159](https://github.com/Accenture/reactive-interaction-gateway/issues/159)
```
[the remark this answers is not in the document]
Comment (jakobwgnr) [reply CHANGELOG.md]: fixed

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,9 +9,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- The proxy configuration can now also be passed as a JSON string. This allows to run the Docker image in environments where mounting a file in a container is not possible. [#159](https://github.com/Accenture/reactive-interaction-gateway/issues/159)
+- When using the proxy, RIG will now add additional forward handlers "X-Content-Type-Options" and "Forwarded" [#113](https://github.com/Accenture/reactive-interaction-gateway/issues/113)
```
[the remark this answers is not in the document]
Comment (jakobwgnr) [reply CHANGELOG.md]: fixed

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,9 +9,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - Increased length of header value in HTTP requests to 16384. Reason is to support long tokens such as SAML.
 
 ### Changed
+- The proxy configuration can now also be passed as a JSON string. This allows to run the Docker image in environments where mounting a file in a container is not possible. [#159](https://github.com/Accenture/reactive-interaction-gateway/issues/159)
+- When using the proxy, RIG will now add additional forward handlers "X-Content-Type-Options" and "Forwarded" [#113](https://github.com/Accenture/reactive-interaction-gateway/issues/113)
```
[the remark this answers is not in the document]
Comment (jakobwgnr) [reply CHANGELOG.md]: hmm - failed on rebasing.... sorry

File: apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex [hunk 5]
```diff
@@ -88,4 +92,36 @@ defmodule RigInboundGateway.ApiProxy.Router do
...
+  # ---
+
+  def add_forward_headers(conn) do
+    %{
+      req_headers: req_headers,
+      remote_ip: remote_ip
+    } = conn
+
+    # GET HOST IP ADDRESS AS STRING
+    {:ok, host_ip} = @host |> String.to_charlist() |> :inet.getaddr(:inet)
```
[the remark this answers is not in the document]
Comment (jakobwgnr) [reply apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex]: unfortunately it didn't work like that... The resolve_addr couldn't be found. I just committed the way on calling it within the method.
I will check how to implement

File: apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex [hunk 6]
```diff
@@ -88,4 +92,36 @@ defmodule RigInboundGateway.ApiProxy.Router do
...
+    host_ip_str = host_ip |> :inet.ntoa() |> to_string
+
+    # GET REMOTE IP AS STRING
+    remote_ip_str = remote_ip |> :inet.ntoa() |> to_string
+
+    forward_headers = [
+      {"X-Content-Type-Options", "nosniff"},
```
[the remark this answers is not in the document]
Comment (kevinbader) [reply apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex]: Sorry - I still don't quite understand why we set this here ^^

According to the MDN docs, [X-Content-Type-Options](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options) is a _response_ header; yet `add_forward_headers` adds it to the _request_ towards the back-end service. Besides, the headers [shouldn't be set unconditionally for all responses](https://github.com/whatwg/fetch/issues/395).

[the remark this answers is not in the document]
Comment (jakobwgnr) [reply apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex]: I'll remove that header

File: apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex [hunk 7]
```diff
@@ -88,4 +92,37 @@ defmodule RigInboundGateway.ApiProxy.Router do
...
+
+    forward_headers = [
+      {"Forwarded", "for=#{remote_ip};by=#{host_ip}"}
+    ]
+
+    updated_headers =
+      for(
+        {key, val} when key not in ["Forwarded"] <- conn.req_headers,
+        do: {key, val}
+      ) ++ forward_headers
```
Comment (kevinbader) [inline apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex]: Hm just noticed the casing might be wrong - [according to the docs](https://hexdocs.pm/plug/Plug.Conn.html):

> req_headers - the request headers as a list, example: [{"content-type", "text/plain"}]. **Note all headers will be downcased** 

Also, the loop can now be written in a simpler way:

```suggestion
    updated_headers =
      conn.req_headers
      |> Enum.reject(fn {k, _} -> k === "forwarded" end)
      |> Enum.concat(forward_headers)
```

I think this also makes a bit clearer what the code is supposed to do.

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- When using the proxy, RIG will now add additional forward header "Forwarded" [#113](https://github.com/Accenture/reactive-interaction-gateway/issues/113)
```
Comment (kevinbader) [inline CHANGELOG.md]:
```suggestion
- When using the proxy, RIG will now add an additional [`Forwarded` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded). [#113](https://github.com/Accenture/reactive-interaction-gateway/issues/113)
```

Comment (jakobwgnr) [reply CHANGELOG.md]: fixed

File: apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex (hunk 7 above)
Comment (jakobwgnr) [reply apps/rig_inbound_gateway/lib/rig_inbound_gateway/api_proxy/router.ex]: fixed

File: apps/rig_inbound_gateway/test/rig/api_proxy/router_test.exs [hunk 9]
```diff
@@ -174,6 +174,17 @@ defmodule RigInboundGateway.ApiProxy.RouterTest do
...
   end
 
+  test_with_server "Any request should include forward headers", @env do
+    route("/myapi/direct", Response.ok!(~s<{"status":"ok"}>))
+
+    conn = call(Router, build_conn(:get, "/myapi/direct"))
+
+    assert conn.status == 200
+    assert conn.resp_body =~ "{\"status\":\"ok\"}"
+    assert [forwarded_header] = get_req_header(conn, "Forwarded")
```
Comment (kevinbader) [inline apps/rig_inbound_gateway/test/rig/api_proxy/router_test.exs]: Fails:
```
  1) test Any request should include forward headers (RigInboundGateway.ApiProxy.RouterTest)
     test/rig/api_proxy/router_test.exs:177
     match (=) failed
     code:  assert [forwarded_header] = get_req_header(conn, "Forwarded")
     right: []
     stacktrace:
       test/rig/api_proxy/router_test.exs:184: (test)
```

Comment (jakobwgnr) [reply apps/rig_inbound_gateway/test/rig/api_proxy/router_test.exs]: "forwarded" was changed to small letter. missed to reflect this in the test..

**Pull request merged**